dali/operators/reader/loader/numpy_loader.cc

```diff
@@ -123,6 +127,86 @@ std::unique_ptr<FileStream> NumpyLoader::ParseHeader(std::unique_ptr<FileStream>
   return file;
 }
 
+// sanitize slabs
+void NumpyLoader::SetupSlab(TensorShape<>& slab_anchor,
+                            TensorShape<>& slab_shape,
+                            const TensorShape<>& sample_shape,
+                            const bool& fortran_order) {
+  int ndims = sample_shape.size();
+
+  if (fortran_order) {
+    // transpose the shapes:
+    TensorShape<> old_anchor(slab_anchor);
+    TensorShape<> old_shape(slab_shape);
+    for (int i = 0; i < ndims; ++i) {
+      slab_anchor[i] = old_anchor[ndims-i-1];
+      slab_shape[i] = old_shape[ndims-i-1];
+    }
+  }
+
+  DALI_ENFORCE((slab_anchor.size() == ndims) &&
+               (slab_shape.size() == ndims),
+               "The dimensions of anchor and slab have to match the sample dims.");
+
+  for (int i = 0; i < ndims; ++i) {
+    int offset = slab_anchor[i];
+    DALI_ENFORCE((offset >= 0) &&
+                 (offset + slab_shape[i] <= sample_shape[i]),
+                 "The slab has to fit inside the sample dimensions");
+  }
+}
+
+std::unique_ptr<FileStream> NumpyLoader::ReadSampleHelper(std::unique_ptr<FileStream> file,
+                                                          ImageFileWrapper& imfile,
+                                                          const NumpyParseTarget& target) {
+  Index image_bytes = target.nbytes();
+  if (copy_read_data_) {
+    if (imfile.image.shares_data()) {
+      imfile.image.Reset();
+    }
+    imfile.image.Resize(target.shape, target.type_info);
+    // copy the image
+    file->Read(static_cast<uint8_t*>(imfile.image.raw_mutable_data()), image_bytes);
+  } else {
+    auto p = file->Get(image_bytes);
+    // Wrap the raw data in the Tensor object.
+    imfile.image.ShareData(p, image_bytes, {image_bytes});
+    imfile.image.Resize(target.shape, target.type_info);
+  }
+  return file;
+}
+
+// here, slab_anchor and slab_shape are passed by value deliberately, because
+// we might need to transpose them. In that case, we do not want to propagate that back
+// to the class members
+std::unique_ptr<FileStream> NumpyLoader::ReadSampleSlabHelper(std::unique_ptr<FileStream> file,
+                                                              ImageFileWrapper& imfile,
+                                                              const NumpyParseTarget& target) {
+  // deal with potential fortran complications
+  TensorShape<> slab_anchor(slab_anchor_);
+  TensorShape<> slab_shape(slab_shape_);
+  SetupSlab(slab_anchor, slab_shape, TensorShape<>(target.shape), target.fortran_order);
+
+  Index image_bytes = target.nbytes();
+  if (copy_read_data_) {
```

Comment:
In this case, we copy anyway. So I think you can forget about `if (copy_read_data_) {` code path anyway.
The idea behind `if (copy_read_data_)` is to make sure we are not keeping too many mmaped files at the same time. If so then we need to copy and allow it to be unmaped just after that. In this case, we copy anyway,  and this tmp `Tensor<CPUBackend> tmp tensor;` tensor in the `else` code path goes out of scope.

Reply:
But there is still a difference between: only reading the relevant parts from a file compared to mapping everything and then slicing the parts in memory. I think we should keep both around, because they might perform differently on internal and external storage. What do you think?

Reply:
Sure.